src/command/pin.rs

```diff
@@ -3,42 +3,40 @@ use structopt::StructOpt;
 use notion_core::error::ErrorDetails;
 use notion_core::session::{ActivityKind, Session};
 use notion_core::tool::ToolSpec;
-use notion_core::version::VersionSpec;
 use notion_fail::{throw, ExitCode, Fallible};
 
 use crate::command::Command;
 
 #[derive(StructOpt)]
 pub(crate) struct Pin {
-    /// The tool to install, e.g. `node` or `npm` or `yarn`
-    tool: String,
-
-    /// The version of the tool to install, e.g. `1.2.3` or `latest`
-    version: Option<String>,
+    /// Tools to pin, like `node@lts` or `yarn@^1.14`.
+    #[structopt(
+        name = "tool[@version]",
+        required = true,
+        min_values = 1,
+        parse(try_from_str = "ToolSpec::try_from_str")
+    )]
+    tools: Vec<ToolSpec>,
 }
 
 impl Command for Pin {
-    fn run(self, session: &mut Session) -> Fallible<ExitCode> {
-        session.add_event_start(ActivityKind::Pin);
-
-        let version = match self.version {
-            Some(version_string) => VersionSpec::parse(&version_string)?,
-            None => VersionSpec::default(),
-        };
-
-        let tool = ToolSpec::from_str_and_version(&self.tool, version);
-
-        match tool {
-            ToolSpec::Node(version) => session.pin_node(&version)?,
-            ToolSpec::Yarn(version) => session.pin_yarn(&version)?,
-            // ISSUE(#292): Implement install for npm
-            ToolSpec::Npm(_version) => unimplemented!("Pinning npm is not supported yet"),
-            ToolSpec::Package(name, _version) => {
-                throw!(ErrorDetails::CannotPinPackage { package: name })
+    fn run(mut self, session: &mut Session) -> Fallible<ExitCode> {
+        self.tools.sort();
+
+        for tool in self.tools {
+            session.add_event_start(ActivityKind::Pin);
```

Comment:
Sorry, misclicked and accidentally deleted your comment, instead of the one I had written 😱 

The question was whether we should do this inside the loop or once per invocation of `run`.

I don't think it matters too much at the moment, the event reporting system isn't really used and hasn't been particularly maintained. I would lean towards once per invocation of `run`, mostly because of the inclusion of an ExitCode in the call to `add_event_end`. Then when we flesh out the event reporting we can add additional `ActivityKind`s to represent the intermediate steps (and likely refactor the events to not always require an exit code to end).

Reply:
Sounds good; I will push it to the outer part of the `run` commands!